sdk/opendp/smartnoise/sql/private_rewriter.py

```diff
@@ -150,32 +172,32 @@ def query(self, query):
                 for ne in query.select.namedExpressions
             ]
         )
+
         select = Select(query.select.quantifier, select)
 
         subquery = Query(
             child_scope.select(), query.source, query.where, query.agg, None, None, None
         )
         subquery = self.exact_aggregates(subquery)
-        subquery = [AliasedRelation(subquery, "exact_aggregates")]
-
+        subquery = [Relation(AliasedSubquery(subquery, Identifier("exact_aggregates")), None)]
         q = Query(select, From(subquery), None, query.agg, query.having, query.order, query.limit)
-
-        return QueryParser(self.metadata).query(str(q))
+        q.load_symbols(self.metadata)
```

Comment:
Can load_symbols be part of the constructor? It seems off to me to have a mutation called after the constructor before returning. @joshua-oss  might have context here

Reply:
I will need to think about this for a bit.  The previous code was round-tripping through the parser, but the new code is more in line with what the original design intention was.  There is actually a note about this in the dev design document:

> The original intention was that AST parsing would branch based on presence of metadata, such that symbol resolution would be a separate and optional step. However, this code path is not well-factored or well-tested, and the current assumption is that metadata is always passed in to parser, and symbol resolution always happens automatically at parse.

So, the original intent was that ASTs can be constructed arbitrarily, and then symbol resolution can happen independently (e.g. clone an AST to multiple backend engines and do symbol resolution for each).  I switched to using `load_symbols` only internally, with external code required to pass through the parser, because I don't have a ton of confidence that the `load_symbols()` code path can handle everything an external party would throw at it.  As long as we are operating only on ASTs that we constructed ourselves, with unit tests, it's safe.  But I can imagine arbitrary externally-constructed ASTs that will blow it up in unexpected ways.

In this case, it may be OK, since we own the rewriter code, and we can be sure unit tests are passing.  The rewriter consumes ASTs that have already been through multiple stages of validation.

Reply:
ok, I understand. 
So what dio you want me to do about it ?

Reply:
Could `metadata` be an optional argument of the `Query` constructor ? In that case, that would change nothing. 
Otherwise, if the AST blocks had type hints, then an external party would not be able to throw anything at it. What do you think ?

Reply:
@vserret, that sounds like a good idea.  Adding it to Query constructor would be an easy change that would let us make `load_symbols` more explicitly internal.  There is still the risk that people could modify the AST after construction, which would cause the symbols to be out of date.  But I think it's OK for that scenario to be out of scope for now.